Python/ceval_gil.c

```diff
@@ -206,58 +206,56 @@ static void recreate_gil(struct _gil_runtime_state *gil)
 #endif
 
 static void
-drop_gil_impl(struct _gil_runtime_state *gil)
-{
-    MUTEX_LOCK(gil->mutex);
-    _Py_ANNOTATE_RWLOCK_RELEASED(&gil->locked, /*is_write=*/1);
-    _Py_atomic_store_int_relaxed(&gil->locked, 0);
-    COND_SIGNAL(gil->cond);
-    MUTEX_UNLOCK(gil->mutex);
-}
-
-static void
-drop_gil(PyInterpreterState *interp, PyThreadState *tstate)
+drop_gil(PyInterpreterState *interp, PyThreadState *tstate, int thread_dying)
 {
     struct _ceval_state *ceval = &interp->ceval;
-    /* If tstate is NULL, the caller is indicating that we're releasing
+    /* If thread_dying is true, the caller is indicating that we're releasing
        the GIL for the last time in this thread.  This is particularly
        relevant when the current thread state is finalizing or its
        interpreter is finalizing (either may be in an inconsistent
        state).  In that case the current thread will definitely
        never try to acquire the GIL again. */
     // XXX It may be more correct to check tstate->_status.finalizing.
-    // XXX assert(tstate == NULL || !tstate->_status.cleared);
+    // XXX assert(thread_dying || !tstate->_status.cleared);
 
+    assert(thread_dying || tstate != NULL);
```

Comment:
I would expect `thread_dying` to directly imply `tstate` is NULL:

```suggestion
    assert(thread_dying == (tstate == NULL));
```

Have I misunderstood?

Reply:
> I would expect `thread_dying` to directly imply `tstate` is NULL

Nope - I'll reply in more detail on your other comment.